File: 0020-isValid/cpp/Solution.cpp
#include <iostream>
#include <stack>
using namespace std;

class Solution
{
public:
    bool isValid(string s)
    {
        std::stack<char> *stack = new std::stack<char>();
        for (char &c : s)
        {
            if (c == '(' || c == '[' || c == '{')
            {
                stack->push(c);
            }
            else
            {
                if (stack->empty())
                {
                    return false;
                }
                char top = stack->top();
                if (top == '(' && c == ')')
                {
                    stack->pop();
                }
                else if (top == '[' && c == ']')
                {
                    stack->pop();
                }
                else if (top == '{' && c == '}')
                {
                    stack->pop();
                }
                else
                {
                    return false;
                }
            }
        }
        return stack->empty();
    }
};

int main(int argc, char const *argv[])
{
    /* code */
    Solution s;
    std::cout << s.isValid("()") << std::endl;
    std::cout << s.isValid("()[]{}") << std::endl;
    std::cout << s.isValid("(]") << std::endl;
    std::cout << s.isValid("([)]") << std::endl;
    std::cout << s.isValid("{[]}") << std::endl;

    return 0;
}

File: 0020-isValid/python/Solution.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        for i in s:
            if i == '(' or i == '[' or i == '{':
                stack.append(i)
            else:
                if len(stack) == 0:
                    return False
                if i == ')' and stack[-1] != '(':
                    return False
                if i == ']' and stack[-1] != '[':
                    return False
                if i == '}' and stack[-1] != '{':
                    return False
                stack.pop()
        return len(stack) == 0

if __name__ == "__main__":
    print(Solution().isValid("()"))
    print(Solution().isValid("()[]{}"))
    print(Solution().isValid("(]"))
    print(Solution().isValid("([)]"))
    print(Solution().isValid("{[]}"))
    print(Solution().isValid(""))
    print(Solution().isValid("(()"))
    print(Solution().isValid("(()[]{})"))
